cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/BasicWorkerContext.java

```diff
@@ -101,7 +102,8 @@ public BasicWorkerContext(WorkerSpecification spec, Program program, RunId runId
     this.transactionSystemClient = transactionSystemClient;
     this.datasetFramework = datasetFramework;
     this.userMetrics = new ProgramUserMetrics(getMetricCollector(metricsCollectionService, program,
-                                                                 runId.getId(), instanceId));
+                                                                 runId.getId(), instanceId,
+                                                                 adapterSpec != null ? adapterSpec.getName() : null));
```

Comment:
since you're calling the ``getName()`` multiple times in this class, is it better to cache it (store it in a field)? Then you wouldn't need this ternary operator - you just pass the field to all these methods, even if it is null.